lib/srv/db/server.go

```diff
@@ -689,7 +699,36 @@ func (s *Server) handleConnection(ctx context.Context, conn net.Conn) error {
 		return trace.Wrap(err)
 	}
 
-	err = engine.HandleConnection(ctx, sessionCtx, conn)
+	if err := engine.InitializeConnection(clientConn, sessionCtx); err != nil {
```

Comment:
Is this separate InitializeConnection method needed? Can't we set clientConn to the engine e.g. in `dispatch` method when creating an engine? Or is it needed to perform some initialization actions (e.g. handleStartup for Postgres), otherwise error won't be sent properly?

Reply:
@r0mant Answering your comments. `InitializeConnection()` is needed (mainly for postgres) to initialize connection with a client otherwise the communication blocks as client and DB agent tries to send data at the same time. Because of that we cannot send an error that comes from `InitializeConnection()` as connection is not ready. 
I think that the best what I can do is to "merge" `InitializeConnection()` into `dispatch()`, so after an engine is created, the connection is ready and we're able to send a message without calling an extra function. Disadvantage is that new engines must follow the same rule, otherwise we may try to send a message when a connection is not ready.